### ISSUE TITLE
The problem of stationary vehicle and how to improve training speed

### ISSUE THREAD

**Issue by zzlqwq** [gerkone/pyTORCS-docker#8]
I am running the ddpg example. The first two Episode car can run the track normally. It seems that they have loaded the trained parameters, but after the third Episode starts, the car has been stopped all time. The terminal is outputting

1/1 [==============================] - 0s 85ms/step
1/1 [==============================] - 0s 12ms/step
1/1 [==============================] - 0s 13ms/step
1/1 [==============================] - 0s 13ms/step
1/1 [==============================] - 0s 13ms/step

I'm trying to read the code to find the problem

**Comment by gerkone** [comment]
Hey. Could you share some more details? Does it show anything else? Are you sure it just stops, or is it just training? Are you using tf2rl DDPG or my version? I don't remember my version having those loading bars but I could be wrong.
In any case both implementations are getting a bit outdated at this point since they still use tensorflow. You could also try with newer agents and see.
Let me know

**Comment by zzlqwq** [comment]
I am using your ddpg not tf2rl DDPG. The complete output is as follows.

[INFO]: Loaded saved actor models
[INFO]: Loaded saved critic models

--------------------------------------------------------------------------

[INFO]: Starting 3000 episodes on track g-track-1
[INFO]: Episode 1/3000 started
[INFO]: Iteration 0 --> Duration 41537.48 ms. Score 353.54. Running average 353.54
[INFO]: Episode 2/3000 started
[INFO]: Iteration 1 --> Duration 62016.21 ms. Score 561.38. Running average 457.46
[INFO]: Starting training: 4 epochs over 2488 collected steps
[INFO]: Completed 4 epochs. Duration 3751.46 ms. Average loss -74.656
[INFO]: Saving models...

--------------------------------------------------------------------------

[INFO]: Episode 3/3000 started
1/1 [==============================] - 0s 126ms/step
1/1 [==============================] - 0s 19ms/step
1/1 [==============================] - 0s 18ms/step
1/1 [==============================] - 0s 13ms/step
1/1 [==============================] - 0s 12ms/step
1/1 [==============================] - 0s 13ms/step
1/1 [==============================] - 0s 13ms/step
1/1 [==============================] - 0s 19ms/step
1/1 [==============================] - 0s 19ms/step
1/1 [==============================] - 0s 18ms/step
1/1 [==============================] - 0s 18ms/step
1/1 [==============================] - 0s 15ms/step
1/1 [==============================] - 0s 14ms/step
1/1 [==============================] - 0s 14ms/step
1/1 [==============================] - 0s 13ms/step
1/1 [==============================] - 0s 19ms/step
1/1 [==============================] - 0s 18ms/step
1/1 [==============================] - 0s 15ms/step
1/1 [==============================] - 0s 12ms/step
1/1 [==============================] - 0s 18ms/step
1/1 [==============================] - 0s 18ms/step
1/1 [==============================] - 0s 19ms/step
1/1 [==============================] - 0s 19ms/step
1/1 [==============================] - 0s 19ms/step
1/1 [==============================] - 0s 19ms/step
1/1 [==============================] - 0s 18ms/step
1/1 [==============================] - 0s 16ms/step
1/1 [==============================] - 0s 16ms/step


It seems to be in training.  In the first two episodes, the agent seems to have loaded the trained network, and it can complete the race well, but after the third episodes, the car don't take any action.I will rewrite the rl algorithm with pythoch, but I'm not sure whether the problem lies in the rl algorithm or the part that interacts with the simulation environment.

**Comment by gerkone** [comment]
If you see the car driving properly in the first couple of episodes it's probably due the warmup phase, where a PID drives (`simple_controller`) instead of the RL agent
https://github.com/gerkone/pyTORCS-docker/blob/d9c01a66c73f9637d93d20151ce8dc3cefd1b176/driver/agents/ddpg/ddpg.py#L120-L126

As for the loading bars honestly I am not sure. It might be that tensorflow at some point introduced loading bars to the `predict` function. Try something like `action = self.actor.model.predict(state, verbose=0)[0]` (line 123 of [ddpg.py](https://github.com/gerkone/pyTORCS-docker/blob/master/driver/agents/ddpg/ddpg.py)).

Other than that I think it's working as intended, you have to give it some time to train, and DDPG is not guaranteed to actually learn anything in such a complex environment with just dense networks.

Let me know if this helps

**Comment by zzlqwq** [comment]
I'm trying to use the ppo algorithm in tf2rl to train. I commented out the parameter(model_dir) in the sim_tf2rl.yaml file, so I haven't loaded any trained network parameters. Its print information is as follows

01:18:20.822 [INFO] (on_policy_trainer.py:131) Total Epi:     1 Steps:    1002 Episode Steps:  1002 Return:  0.2713 TIME(s):  44.90 FPS: 22.32
01:19:04.022 [INFO] (on_policy_trainer.py:131) Total Epi:     2 Steps:    2004 Episode Steps:  1002 Return:  0.3485 TIME(s):  43.19 FPS: 23.20
01:19:47.227 [INFO] (on_policy_trainer.py:131) Total Epi:     3 Steps:    3006 Episode Steps:  1002 Return:  0.3405 TIME(s):  43.20 FPS: 23.19
01:20:36.362 [INFO] (on_policy_trainer.py:206) Done 12 epochs. Average actor loss 0.004451490938663483
01:20:37.734 [INFO] (on_policy_trainer.py:131) Total Epi:     4 Steps:    4008 Episode Steps:  1002 Return:  0.2966 TIME(s):  50.50 FPS: 19.84
01:21:20.925 [INFO] (on_policy_trainer.py:131) Total Epi:     5 Steps:    5010 Episode Steps:  1002 Return:  0.2362 TIME(s):  43.19 FPS: 23.20
01:22:04.132 [INFO] (on_policy_trainer.py:131) Total Epi:     6 Steps:    6012 Episode Steps:  1002 Return:  0.2622 TIME(s):  43.21 FPS: 23.19
01:22:47.328 [INFO] (on_policy_trainer.py:131) Total Epi:     7 Steps:    7014 Episode Steps:  1002 Return:  0.2362 TIME(s):  43.19 FPS: 23.20
01:23:34.708 [INFO] (on_policy_trainer.py:206) Done 12 epochs. Average actor loss 0.0024660788476467133
01:23:36.409 [INFO] (on_policy_trainer.py:131) Total Epi:     8 Steps:    8016 Episode Steps:  1002 Return:  0.2943 TIME(s):  49.08 FPS: 20.42
01:24:19.587 [INFO] (on_policy_trainer.py:131) Total Epi:     9 Steps:    9018 Episode Steps:  1002 Return:  0.3534 TIME(s):  43.18 FPS: 23.21

It seems to be in normal training, but in torcs, it seems that the vehicle hardly moves (Reward is also very small). This should be the same problem I encountered when I used your version of ddpg. 

When I open the instrument cluster, I see that the throttle value of the vehicle changes from 0 to 1, but the steering information changes normally. Is it possible that the output value of the network is not restored after normalization?

But when I load the ppo network parameters that you have trained, it can work normally, which is very strange.

**Comment by zzlqwq** [comment]
Yes, you are right. It takes a while for the vehicle to learn to move forward. I have one last question. I want to know whether I can speed up the training process. I noticed that during my training, the fps is around 20, which does not occupy much of my computing resources. Do I have a way to speed up training? Thank you.

**Comment by gerkone** [comment]
Hey. Good to hear that.
About the performance at the moment I really don't have a solution. I know it's quite slow and performance has always been an issue. Fps is capped and as far as I remember if you raise the limit it messes up the physics. Also I don't know of any alternative torcs environment that does that (maybe [madras](https://github.com/madras-simulator/MADRaS) if you have a distributed algorithm, but I never got it to work that great in general).
I was planning to rewrite torcs as a pybind library, remove the window and so on but never got the time to do it.

Also, could you change the title to something more fitting to this discussion?
Let me know if I can help on anything else, thanks

**Comment by zzlqwq** [comment]
OK, thank you very much for your work and reply. I have no other questions.
